### PR TITLE
ENH: handler.otp stores OTP in storelocations. Fixes #580

### LINKED ISSUE #580: handler.otp() does not save in storelocations.otp
`handler.otp(...)` stores data in the **session** store, not `storelocations.otp` as indicated by https://gramener.com/gramex/guide/auth/#otp

Raised by @nikhilkabbin

### PR DESCRIPTION
**Pull request by sanand0**

### CODE REVIEW

File: gramex/handlers/basehandler.py [hunk 1]
```diff
@@ -692,27 +694,25 @@ def save_session(self):
         if getattr(self, '_session', None) is not None:
             self._session_store.dump(self._session['id'], self._session)
 
-    def otp(self, expire=60, prefix='otp', user=None):
+    def otp(self, expire=60, user=None, email='OTP'):
```
Comment (bhatsandeep) [inline gramex/handlers/basehandler.py:697]: email=OTP is unclear

File: gramex/services/storelocations.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+        - `type` (str), e.g. `"VARCHAR(10)"`
+        - `default` (str/int/float/bool), e.g. `"none@example.org"`
+        - `nullable` (bool), e.g. `False`
+        - `primary_key` (bool), e.g. `True` -- used only when creating new tables
+        - `autoincrement` (bool), e.g. `True` -- used only when creating new tables
+
+        The parameters are the same as [gramex.data.alter][].
+        '''
+        self.store = {'url': url, 'table': table, 'columns': columns, **kwargs}
```
Comment (bhatsandeep) [inline gramex/services/storelocations.py:32]: Need to document that the store structure has to be adhered to irrespective of what storage location is used.

File: gramex/services/storelocations.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        Returns:
+            Dict with user object, email, expire and token. None if no match
+        '''
+        rows = gramex.data.filter(**self.store, args={'token': [token]})
+        if len(rows) > 0:
+            row = rows.iloc[0].to_dict()
+            row['user'] = json.loads(row['user'])
+            if row['expire'] > time.time():
+                return row
+        return None
```
Comment (bhatsandeep) [inline gramex/services/storelocations.py:95]: Would need a purge option for tokens that have expired.

**Comment by radheyakale** [comment]
Do we need to write test cases for all storage locations or is that already done?

**Comment by sanand0** [comment]
> Do we need to write test cases for all stores or is that already done?

The test cases for pipeline logging and OTP cover both kinds of storelocations -- so we have full coverage on this.

File: gramex/handlers/basehandler.py (hunk 1 above)
Comment (sanand0) [reply gramex/handlers/basehandler.py]: Changed it to `type='OTP'`

File: gramex/services/storelocations.py (hunk 2 above)
Comment (sanand0) [reply gramex/services/storelocations.py]: Documented this.

File: gramex/services/storelocations.py (hunk 3 above)
Comment (sanand0) [reply gramex/services/storelocations.py]: Done

**Pull request merged**